flax/testing/benchmark.py

```diff
@@ -0,0 +1,239 @@
+  assert methods for checking benchmark target metrics. This is also
+  necessary for correctly reporting benchmark results and determining its
+  success.
+  """
+
+  def __init__(self, *args, **kwargs):
+    """Wrap test methods in a try-except decorator to delay exceptions."""
+    super(Benchmark, self).__init__(*args, **kwargs)
+    for func_name in dir(self):
+      if func_name.startswith('assert'):
+        func = getattr(self, func_name)
+        patched_func = functools.partial(
+            self._collect_assert_wrapper, func=func)
+        setattr(self, func_name, patched_func)
+
+  # pylint: disable=invalid-name
+  def _collect_assert_wrapper(self, *args, func=None, **kwargs):
+    """Wrapper around assert methods that caputres and collects failures."""
+    try:
+      return func(*args, **kwargs)
+    except self.failureException as ex:
+      self._outstanding_fails.append(ex)
+
+  def setUp(self):
+    """Setup ran before each test."""
+    super(Benchmark, self).setUp()
+    self._reported_name = None
+    self._reported_wall_time = None
+    self._reported_metrics = {}
+    self._reported_extras = {}
+    self._outstanding_fails = []
+
+  def tearDown(self):
+    """Tear down after each test."""
+    super(Benchmark, self).tearDown()
+    self._report_benchmark_results()
+    for message in self._outstanding_fails:
+      raise self.failureException(message)
+
+  def has_outstanding_fails(self):
+    """Determine whether the benchmark failed, but the error is deferred."""
+    return len(self._outstanding_fails) > 0
+
+  def read_summaries(self, path):
+    """Read TensorBoard summaries."""
+    return _get_tensorboard_scalars(path)
+
+  def report_wall_time(self, wall_time: float):
+    """Report wall time for the benchmark."""
+    self._update_reported_name()
+    self._reported_wall_time = wall_time
+
+  def report_metrics(self, metrics: Dict[str, float]):
+    """Report metrics for the benchmark."""
+    self._update_reported_name()
+    self._reported_metrics.update(metrics)
+
+  def report_metric(self, name: str, value: float):
+    """Report a single metric for the benchmark."""
+    self.report_metrics({name: value})
+
+  def report_extras(self, extras: Dict[str, str]):
+    """Report extras for the benchmark."""
+    self._update_reported_name()
+    self._reported_extras.update(extras)
+
+  def report_extra(self, name: str, value: str):
+    """Report a single extra for the benchmark."""
+    self.report_extras({name: value})
+
+  def _get_test_name(self, prefix='test_'):
+    """Returns full name of test class and method calling report_benchmark.
+
+    The name is based on the *outermost* Benchmark class in the class stack.
+    Based on tensorflow/python/platform/benchmark.py
+
+    Args:
+      prefix: str. Prefix that the caller method must have.
+
+    Returns:
+      Resolved test name as `ClassName.test_name`.
+    """
+    # Find the caller method (outermost Benchmark class).
+    stack = inspect.stack()
```

Comment:
Interesting approach, it's great if it works!

Reply:
It seems to, at least in the examples I've tried. Personally, I wasn't too happy with the solution, but couldn't think of anything nicer.